### PR TITLE
[multi-stage] Support query plan for GROUP BY only in the intermediary stage

### PR DESCRIPTION
**Pull request by vvivekiyer**
Currently, when we build a `StagePlan` for GROUP BY aggregations, we create a Leaf Level Aggregation and an Intermediate Level Aggregation. 

In certain cases, for high cardinality columns, doing partial aggregation at leaf level might not be useful to collapse data by a considerable degree. If the optimizer can recognize this fact (based on SQL Hint or cardinality information), we can omit the aggregation at leaf stage and just do it at the intermediate stage. 

```
Default / existing GROUP BY plan in multi stage engine

- scan / filter / project + partial group by aggregation at leaf
- hash exchange
- full aggregates at intermediary
```

```
For high cardinality GROUP BY, a different plan can be tried potentially to improve performance 

- scan / filter / project at leaf
- hash exchange
- group by aggregation at intermediary
```

This optimization will be turned off by default. The user can force this optimization by using the SQL Hint "SKIP_LEAF_STAGE_GROUP_BY_AGGREGATION"

Other changes in this PR:
1. We currently use regex to strip out all comments from queries. This code was added for https://github.com/apache/pinot/issues/7714 when we had the older option support. As this is no longer the case, I've removed this code and associated tests.
2. Because of (1), it opens up the opportunity for us to support `SQLHints` in the multistage engine. The advantage over options is that, SQLHints will be automatically propagated down to the required RelNode by Calcite. I've added some code to support SQL Hints with this PR.
3. Some variable naming cleanup.

### CODE REVIEW

**Comment by siddharthteotia** [comment]
Some more info - 

We probably don't want this to be the default plan for GROUP BY in multi stage. Default should be the existing 2 phase GROUP BY (leaf + exchange + intermediary).

Just want support (preferably via hint where we know what we are doing) for this to be there for testing / tuning high cardinality GROUP BY on multi stage engine. There could be a performance benefit by avoiding the partial aggregates at the leaf layer for such a scenario.

**Comment by ankitsultana** [comment]
> Default should be the existing 2 phase GROUP BY (leaf + exchange + intermediary).

+1 to leveraging query hints to trigger this and keeping default behavior as is.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 1]
```diff
@@ -91,6 +91,92 @@
           "\n                  LogicalTableScan(table=[[a]])",
           "\n"
         ]
+      },
+      {
+        "description": "SQL hint based group by optimization with select and aggregate column",
```
Comment (somandal) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:96]: From my understanding of the code this hint can potentially be used for the following types of queries in addition to the group by queries tested here:

- Select only aggregation columns with group-by (e.g. `SELECT SUM(a.col3) FROM a GROUP BY a.col3`)
- Select only aggregation columns without group by (e.g. `SELECT SUM(a.col3) FROM a`)

Do you intend this hint to act on the above type of queries?
If so, can you add some test cases for the above?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 2]
```diff
@@ -91,6 +91,92 @@
...
+          "\n  LogicalExchange(distribution=[hash[2]])",
+          "\n    LogicalProject(col2=[$0], col3=[$1], col1=[$2])",
+          "\n      LogicalFilter(condition=[AND(>=($1, 0), =($0, 'a'))])",
+          "\n        LogicalTableScan(table=[[a]])",
+          "\n"
+        ]
+      },
+      {
+        "description": "SQL hint based group by optimization on 2 columns with filter",
+        "sql": "EXPLAIN PLAN FOR SELECT /*+ SKIP_LEAF_STAGE_GROUP_BY_AGGREGATION */ a.col2, a.col1, SUM(a.col3) FROM a WHERE a.col3 >= 0 AND a.col1 = 'a'  GROUP BY a.col1, a.col2",
```
Comment (somandal) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:136]: Also add a test where you perform multiple aggregations on the same column (e.g. a.col3), ensure that the code path to project the given column (a.col3) only once is working as expected.

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 3]
```diff
@@ -169,36 +183,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    RelNode childRel = ((HepRelVertex) oldAggRel.getInput()).getCurrentRel();
+    LogicalProject project;
+
+    List<Integer> newAggArgColumns = new ArrayList<>();
+    List<Integer> newAggGroupByColumns = new ArrayList<>();
+
+    // 1. Create the LogicalProject node if it does not exist. This is to send only the relevant columns over
+    //    the wire for intermediate aggregation.
+    if (childRel instanceof Project) {
+      // Avoid creating a new LogicalProject if the child node of aggregation is already a project node.
```
Comment (somandal) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:237]: Just curious, can this code path result in projecting the same column multiple times if it's part of multiple aggregation functions?

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 4]
```diff
@@ -119,23 +132,25 @@ public void onMatch(RelOptRuleCall call) {
...
 
     // add the exchange as the input node to the relation builder.
     RelBuilder relBuilder = ruleCall.builder();
     relBuilder.push(exchange);
-    List<RexNode> inputExprs = new ArrayList<>(relBuilder.fields());
 
     // make input ref to the exchange after the leaf aggregate.
     RexBuilder rexBuilder = exchange.getCluster().getRexBuilder();
     final int nGroups = oldAggRel.getGroupCount();
-    for (int i = 0; i < nGroups; i++) {
```
Comment (somandal) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:136]: why was this removed? this sets up the inputrefs for the exchange, right?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 5]
```diff
@@ -91,6 +91,92 @@
           "\n                  LogicalTableScan(table=[[a]])",
           "\n"
         ]
+      },
+      {
+        "description": "SQL hint based group by optimization with select and aggregate column",
+        "sql": "EXPLAIN PLAN FOR SELECT /*+ SKIP_LEAF_STAGE_GROUP_BY_AGGREGATION */ a.col1, SUM(a.col3) FROM a GROUP BY a.col1",
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:97]: Can you add a separate test for GROUP BY with `AVG()` as well ?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 6]
```diff
@@ -91,6 +91,92 @@
...
+      {
+        "description": "SQL hint based group by optimization with having clause",
+        "sql": "EXPLAIN PLAN FOR SELECT /*+ SKIP_LEAF_STAGE_GROUP_BY_AGGREGATION */ a.col1, COUNT(*), SUM(a.col3) FROM a WHERE a.col3 >= 0 AND a.col2 = 'a' GROUP BY a.col1 HAVING COUNT(*) > 10 AND MAX(a.col3) >= 0 AND MIN(a.col3) < 20 AND SUM(a.col3) <= 10 AND AVG(a.col3) = 5",
+        "output": [
+          "Execution Plan",
+          "\nLogicalProject(col1=[$0], EXPR$1=[$1], EXPR$2=[$2])",
+          "\n  LogicalFilter(condition=[AND(>($1, 10), >=($3, 0), <($4, 20), <=($2, 10), =($5, 5))])",
+          "\n    LogicalProject(col1=[$0], EXPR$1=[$1], EXPR$2=[$2], $f3=[$3], $f4=[$4], $f5=[CAST(/($5, $1)):INTEGER NOT NULL])",
+          "\n      LogicalProject(col1=[$0], EXPR$1=[$1], EXPR$2=[$2], $f3=[$3], $f4=[$4], $f5=[$2])",
+          "\n        LogicalAggregate(group=[{2}], EXPR$1=[COUNT()], EXPR$2=[$SUM0($1)], agg#2=[MAX($1)], agg#3=[MIN($1)])",
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:156]: Can you add a test for the scenario where `COUNT(*)` is not used in the aggregation set but `AVG()` is still used in the `HAVING` clause ?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 1 above)
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:96]: Do we need to add corresponding tests for the same queries you have used here but without the hint ? Or do they already exist ?

Comment (siddharthteotia) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json:96]: I don't think this hint is applicable to the 2nd type of queries unless I am missing something.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 7]
```diff
@@ -91,6 +91,92 @@
...
+          "\n          LogicalExchange(distribution=[hash[2]])",
+          "\n            LogicalProject(col2=[$0], col3=[$1], col1=[$2])",
+          "\n              LogicalFilter(condition=[AND(>=($1, 0), =($0, 'a'))])",
+          "\n                LogicalTableScan(table=[[a]])",
+          "\n"
+        ]
+      },
+      {
+        "description": "SQL hint based group by optimization with having clause and select alias",
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:165]: Please consider adding few tests with ORDER BY too.

**Comment by somandal** [comment]
One more question, this PR shouldn't require any further changes on the execution side, right? In that case, can you also add some execution tests (using the output mechanism if H2 doesn't work for this) and compare the output of the query with hint and without hint to make sure the results match?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 1 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json:110]: Good catch. This is needed only for group by aggregates. Fixed it.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 2 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json]: We have this implicit test where we perform multiple aggregations on a single column (some of the aggregations are in HAVING clause).

But added an explicit test as well.

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java (hunk 3 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:241]: No. Calcite project makes sure to project the column only once. I added a test case for this.

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java (hunk 4 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java]: This seemed unnecessary based on my code reading. But I am not sure, so adding it back.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 5 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json]: Added.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 6 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json:210]: Added

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 1 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json:110]: It already exists. Added similar tests for newly added cases as well.

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json (hunk 7 above)
Comment (vvivekiyer) [reply pinot-query-planner/src/test/resources/queries/GroupByPlans.json:236]: Added.

**Comment by vvivekiyer** [comment]
> One more question, this PR shouldn't require any further changes on the execution side, right? In that case, can you also add some execution tests (using the output mechanism if H2 doesn't work for this) and compare the output of the query with hint and without hint to make sure the results match?

@somandal  That's right. No execution side changes are necessary. I've added some runtime tests. Please review.

**Review by somandal — APPROVED** [review]
LGTM!

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintStrategyTable.java [hunk 8]
```diff
@@ -18,19 +18,36 @@
...
 
   public static final HintStrategyTable PINOT_HINT_STRATEGY_TABLE = HintStrategyTable.builder()
       .hintStrategy(INTERNAL_AGG_INTERMEDIATE_STAGE, HintPredicates.AGGREGATE)
       .hintStrategy(INTERNAL_AGG_FINAL_STAGE, HintPredicates.AGGREGATE)
+      .hintStrategy(SKIP_LEAF_STAGE_GROUP_BY_AGGREGATION, HintPredicates.AGGREGATE)
       .build();
+
+  public static boolean containsHint(ImmutableList<RelHint> hintList, String hintName) {
+    for (RelHint relHint : hintList) {
+      if (relHint.hintName.equals(hintName)) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintStrategyTable.java:47]: Should this be `equalsIgnoreCase` ?

Comment (siddharthteotia) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintStrategyTable.java:47]: (nit) not a big deal but if invoked multiple times, may be hintList should be a `Set` I guess

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 9]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    rexBuilder.addAggCall(newCall, nGroups, newCalls, aggCallMapping, oldAggRel.getInput()::fieldIsNullable);
   }
 
   private static List<Integer> convertArgList(int oldCallIndexWithShift, List<Integer> argList) {
     Preconditions.checkArgument(argList.size() <= 1,
         "Unable to convert call as the argList contains more than 1 argument");
     return argList.size() == 1 ? Collections.singletonList(oldCallIndexWithShift) : Collections.emptyList();
   }
+
+  private void createPlanWithoutLeafAggregation(RelOptRuleCall call) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:229]: (nit) suggest renaming this to `createPlanWithoutLeafGroupByAggregation`

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 10]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
     Preconditions.checkState(SUPPORTED_AGG_KIND.contains(aggKind), "Unsupported SQL aggregation "
         + "kind: {}. Only splittable aggregation functions are supported!", aggKind);
 
-    // Special treatment on COUNT
     AggregateCall newCall;
-    if (oldAggregation instanceof SqlCountAggFunction) {
-      newCall = AggregateCall.create(new SqlSumEmptyIsZeroAggFunction(), oldCall.isDistinct(), oldCall.isApproximate(),
-          oldCall.ignoreNulls(), convertArgList(nGroups + oldCallIndex, Collections.singletonList(oldCallIndex)),
-          oldCall.filterArg, oldCall.distinctKeys, oldCall.collation, oldCall.type, oldCall.getName());
+    if (isLeafStageAggregationPresent) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:197]: (nit) It might be cleaner to create a helper function for this code

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 11]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    int projectIndex = 0;
+    for (int group : oldAggRel.getGroupSet().asSet()) {
+      projectColNames.add(fieldList.get(group).getName());
+      projectColRexNodes.add(childRexBuilder.makeInputRef(childRel, group));
+      projectSet.put(group, projectColRexNodes.size() - 1);
+      newAggGroupByCols.add(projectIndex++);
+    }
+
+    List<AggregateCall> oldAggCallList = oldAggRel.getAggCallList();
+    for (int i = 0; i < oldAggCallList.size(); i++) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:281]: (nit) suggest moving `oldAggCallList.size()` once before the loop

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 12]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+      List<Integer> argList = oldAggCallList.get(i).getArgList();
+      if (argList.size() == 0) {
+        newAggArgColumns.add(-1);
+        continue;
+      }
+      for (int j = 0; j < argList.size(); j++) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:287]: same here

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 13]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    RelNode childRel = ((HepRelVertex) oldAggRel.getInput()).getCurrentRel();
+    LogicalProject project;
+
+    List<Integer> newAggArgColumns = new ArrayList<>();
+    List<Integer> newAggGroupByColumns = new ArrayList<>();
+
+    // 1. Create the LogicalProject node if it does not exist. This is to send only the relevant columns over
+    //    the wire for intermediate aggregation.
+    if (childRel instanceof Project) {
+      // Avoid creating a new LogicalProject if the child node of aggregation is already a project node.
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:240]: > Avoid creating a new LogicalProject if the child node of aggregation is already a project node.

Shouldn't this always be the case ?

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 14]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+
+    for (int i = 0; i < oldAggCallList.size(); i++) {
+      if (oldAggCallList.get(i).getArgList().size() == 0) {
+        // This can be true for COUNT. Add a placeholder value which will be ignored.
+        newAggArgColumns.add(-1);
+        continue;
+      }
+      for (int j = 0; j < oldAggCallList.get(i).getArgList().size(); j++) {
+        Integer col = oldAggCallList.get(i).getArgList().get(j);
+        newAggArgColumns.add(col);
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:314]: (nit) suggest making this code more readable / clean. Lots of chained function calls

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 15]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    return LogicalProject.create(childRel, Collections.emptyList(), projectColRexNodes, projectColNames);
+  }
+
+  private List<Integer> fetchNewAggArgCols(List<AggregateCall> oldAggCallList) {
+    List<Integer> newAggArgColumns = new ArrayList<>();
+
+    for (int i = 0; i < oldAggCallList.size(); i++) {
+      if (oldAggCallList.get(i).getArgList().size() == 0) {
+        // This can be true for COUNT. Add a placeholder value which will be ignored.
+        newAggArgColumns.add(-1);
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:309]: (nit) may want to use a private static final constant instead of -1 since couple of places use it

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 16]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    }
+
+    return LogicalProject.create(childRel, Collections.emptyList(), projectColRexNodes, projectColNames);
+  }
+
+  private List<Integer> fetchNewAggArgCols(List<AggregateCall> oldAggCallList) {
+    List<Integer> newAggArgColumns = new ArrayList<>();
+
+    for (int i = 0; i < oldAggCallList.size(); i++) {
+      if (oldAggCallList.get(i).getArgList().size() == 0) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:307]: Do we have unit tests for all scenarios ?

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 17]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+    List<RexNode> projectColRexNodes = new ArrayList<>();
+    List<String> projectColNames = new ArrayList<>();
+    // Maintains a mapping from the column to the corresponding index in projectColRexNodes.
+    Map<Integer, Integer> projectSet = new HashMap<>();
+
+    int projectIndex = 0;
+    for (int group : oldAggRel.getGroupSet().asSet()) {
+      projectColNames.add(fieldList.get(group).getName());
+      projectColRexNodes.add(childRexBuilder.makeInputRef(childRel, group));
+      projectSet.put(group, projectColRexNodes.size() - 1);
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:276]: (nit) we already know the running size.. probably not necessary to do `projectColRexNodes.size() - 1`

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 18]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+
+    // 3. Create an intermediate stage aggregation.
+    RelNode newAggNode =
+        makeNewIntermediateAgg(call, oldAggRel, exchange, false, newAggArgColumns, newAggGroupByColumns);
+
+    call.transformTo(newAggNode);
+  }
+
+  private LogicalProject createLogicalProjectForAggregate(Aggregate oldAggRel, List<Integer> newAggArgColumns,
+      List<Integer> newAggGroupByCols) {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:262]: You may want to name these lists differently since these are tracking column indices essentially

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java [hunk 19]
```diff
@@ -169,36 +185,136 @@ private RelNode makeNewIntermediateAgg(RelOptRuleCall ruleCall, Aggregate oldAgg
...
+      }
+      for (int j = 0; j < argList.size(); j++) {
+        Integer col = argList.get(j);
+        if (!projectSet.containsKey(col)) {
+          projectColRexNodes.add(childRexBuilder.makeInputRef(childRel, col));
+          projectColNames.add(fieldList.get(col).getName());
+          projectSet.put(col, projectColRexNodes.size() - 1);
+          newAggArgColumns.add(projectColRexNodes.size() - 1);
+        } else {
+          newAggArgColumns.add(projectSet.get(col));
```
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:295]: I think it is possible to do `newAggArgColumns.add()` commonly outside the if-else block.

(nit) nbd.. but just seeing how we can make this code slightly more readable / less verbose :)

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java (hunk 11 above)
Comment (siddharthteotia) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateExchangeNodeInsertRule.java:281]: I downloaded the patch and put this in debugger mode to clarify few things about my understanding of the code. It is clear to me now. 

But you may want to consider adding a few comments explaining the code / algorithm and / or give an example in comments for future readers and

File: pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java [hunk 20]
```diff
@@ -109,6 +109,21 @@ protected Object[][] provideQueries() {
...
+        new Object[]{"SELECT /*+ skipLeafStageGroupByAggregation */ a.col2, a.col1, SUM(a.col3) FROM a WHERE a"
+            + ".col3 >= 0 AND a.col1 = 'a'  GROUP BY a.col1, a.col2"},
+        new Object[]{"SELECT /*+ skipLeafStageGroupByAggregation */ a.col1, AVG(b.col3) FROM a JOIN b ON a.col1 "
+            + "= b.col2  WHERE a.col3 >= 0 AND a.col2 = 'a' AND b.col3 < 0 GROUP BY a.col1"},
+        new Object[]{"SELECT /*+ skipLeafStageGroupByAggregation */ a.col1 as v1, a.col1 as v2, AVG(a.col3) FROM"
+            + " a GROUP BY v1, v2"},
+        new Object[]{"SELECT /*+ skipLeafStageGroupByAggregation */ a.col2, COUNT(*), SUM(a.col3), SUM(a.col1) "
+            + "FROM a WHERE a.col3 >= 0 AND a.col2 = 'a' GROUP BY a.col2 HAVING COUNT(*) > 10 AND MAX(a.col3) >= 0 "
+            + "AND MIN(a.col3) < 20 AND SUM(a.col3) <= 10 AND AVG(a.col3) = 5"},
+        new Object[]{"SELECT /*+ skipLeafStageGroupByAggregation */ a.col2, a.col3 FROM a JOIN b ON a.col1 = b"
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/java/org/apache/pinot/query/QueryEnvironmentTestBase.java:125]: I wonder if using this hint for this query is a NO-OP ? I mean in this case, GROUP BY is on top of JOIN, so this particular hint's purpose is not applicable ?

Curious what does the plan look like for this query with and without the hint ?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 21]
```diff
@@ -13,6 +13,20 @@
           "\n"
         ]
       },
+      {
+        "description": "SQL hint based group by optimization with select and multiple aggregations one 1 column",
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:17]: The test query doesn't have the hint. Did you mean to add the hint to this query ?

File: pinot-query-planner/src/test/resources/queries/GroupByPlans.json [hunk 22]
```diff
@@ -13,6 +13,20 @@
           "\n"
         ]
       },
+      {
```
Comment (siddharthteotia) [inline pinot-query-planner/src/test/resources/queries/GroupByPlans.json:16]: Can you also cover that JOIN related test case here where I commented about this hint being a NO-OP ? Just want to see the plan with and without hint

**Review by siddharthteotia — APPROVED** [review]
LGTM overall.

**Comment by siddharthteotia** [comment]
@vvivekiyer - please make sure to address (as needed) the comments in the follow-up PRs
